Added VQA as a evaluator.

ruhanaazam opened the pull request:
VQA will evaluate whether a task is completed given screenshots taken throughout the execution. The evaluator requires that `--take_screenshot true` is set.

teaxio commented on test/evaluators.py:
```diff
@@ -396,6 +400,46 @@ async def __call__(
+        score = -1.0
+
+        # Get path to screenshots for the given task
+        test_folder = list_items_in_folder(f"{os. getcwd()}/test/logs/")[-1] # Get the most recent log folder, this may take look for the wrong folder TODO: fix to take correct folder
```
Reference the test_processor.py on now to get the logs directory. Please use that approach to be consistent with paths.

teaxio commented on test/test_utils.py:
```diff
@@ -261,3 +262,36 @@ def get_formatted_current_timestamp(format: str = "%Y-%m-%d %H:%M:%S") -> str:
+    except PermissionError:
+        return f"Permission denied to access {path}."
+
+def compress_png(file_path, max_size_mb=20, reduce_factor=0.9):
```
please document this method and some inline docs for the compression would be informative to others

ruhanaazam replied:
Added inline documentation.

teaxio commented on test/evaluators.py:
```diff
@@ -9,13 +9,17 @@
 from termcolor import colored
 
+import os
+from .validation_agent.validator import validate_task_vqa
```
Where is this file? It does not seem to be in the PR

ruhanaazam replied:
The validation_agent directory is another github repo. Is there a best practice for how this repo should be added to this project? Some options: 1) Clone the repo and push 2) Add instructions to cloning and setting up the repo to the Readme

ruhanaazam replied:
The final decision was to add the validation_agent directory independently to this repo.

deepak-akkil commented:
Please add a section in readme on VQA (and more generally evaluations). I am thinking a new sub section in the test related readme part. Something like:

### Automatic Evaluation Types:
 We support different evaluation types such as: 
    - `string_match`: compares the text response from Agent-E with a reference answer. There are multiple types of string_match that is supported such as: exact_match | must_include | some_matches | fuzzy_match
    - `url_match`: compares the final URL to the reference url. 
    - `program_html`: Use a JS code to retrieve a value from the HTML page and compare it with a reference answer. 
    - `manual`: pauses the test execution for manual pass/fail annotation.
    - `'vqa`: Uses GPT4-vision to evaluate the task success based on a sequence of screenshots, taken at each step. If you use vqa, you need to have `--take_screenshots true` when running the test.
Examples of each can be found in "test/tasks/test.json"

Ensure that there is atleast one vqa example in test.json

deepak-akkil commented:
Another issue i found when running this version was since we do not ever remove screenshots from previous run, just repeating the same task again takes all screenshots in the snapshot folder. This adds up the cost + also sometimes lead to incorrect vqa evaluation (imagine if in the earlier run the task was successful and in the next run it wasnt).

 Perhaps an easy solution is to clear snapshots folder if a new run is taking place?

deepak-akkil commented:
Otherwise looks good to me!!

deepak-akkil reviewed (commented):
Few minor comments added, otherwise looks good. 
Tested to work fine as well when the snapshot folder is empty

deepak-akkil commented:
Also add some default fall back (may fail the test or keep it unknown) when the snapshots folder is empty. 
Today it crashes the test with the following print statement
`VQA score is None becauase None`

ruhanaazam commented:
@deepak-akkil I addressed your comments:
  1) Improved error handling when exceptions are thrown by VQA: Exceptions are more descriptive now. And VQA evaluation failures are treated as `Skips` so evaluation should continue if there is an issue with the VQA evaluation.
  2) Fixed linter ruff complaints.
  3) Added a warning to users when no screenshots are found for VQA evaluation.
  4) Deleted all screenshots from prior runs before starting a new run (with the same test name).

Let me know if there are any other issues.
